R/model.R

```diff
@@ -1,63 +1,232 @@
+              direct_sample = direct_sample,
+              properties = properties)
   class(ret) <- "mcstate_model"
   ret
 }
+
+
+
+validate_model_properties <- function(properties, call = NULL) {
+  if (is.null(properties)) {
+    return(mcstate_model_properties())
+  }
+
+  if (!inherits(properties, "mcstate_model_properties")) {
+    cli::cli_abort(
+      "Expected 'properties' to be a 'mcstate_model_properties' object",
+      arg = "properties", call = call)
+  }
+
+  properties
+}
+
+
+validate_model_parameters <- function(model, call = NULL) {
+  if (!is.character(model$parameters)) {
+    cli::cli_abort("Expected 'model$parameters' to be a character vector",
+                   arg = "model", call = call)
+  }
+  model$parameters
+}
+
+
+validate_model_domain <- function(model, call = NULL) {
+  n_pars <- length(model$parameters)
+  if (is.null(model$domain)) {
+    domain <- cbind(rep(-Inf, n_pars), rep(Inf, n_pars))
+  } else {
+    domain <- model$domain
+    if (!is.matrix(domain)) {
+      cli::cli_abort("Expected 'model$domain' to be a matrix if non-NULL")
+    }
+    if (nrow(domain) != n_pars) {
+      cli::cli_abort(paste(
+        "Expected 'model$domain' to have {n_pars} row{?s},",
+        "but it had {nrow(domain)}"))
+    }
+    if (ncol(domain) != 2) {
+      cli::cli_abort(paste(
+        "Expected 'model$domain' to have 2 columns,",
+        "but it had {ncol(domain)}"))
+    }
+  }
+  domain
+}
+
+
+validate_model_density <- function(model, call = NULL) {
+  if (!is.function(model$density)) {
+    cli::cli_abort("Expected 'model$density' to be a function",
+                   arg = "model", call = call)
+  }
+  model$density
+}
+
+
+validate_model_gradient <- function(model, properties, call) {
+  if (isFALSE(properties$has_gradient)) {
+    return(NULL)
+  }
+  gradient <- model$gradient
+  if (isTRUE(properties$has_gradient) && !is.function(gradient)) {
+    cli::cli_abort(
+      paste("Did not find a function 'gradient' within your model, but",
+            "your properties say that it should do"),
+      arg = "model", call = call)
+  }
+  if (!is.null(gradient) && !is.function(gradient)) {
+    cli::cli_abort(
+      "Expected 'model$gradient' to be a function if non-NULL",
+      arg = "model", call = call)
+  }
+  gradient
+}
+
+
+validate_model_direct_sample <- function(model, properties, call) {
+  if (isFALSE(properties$has_direct_sample)) {
+    return(NULL)
+  }
+  direct_sample <- model$direct_sample
+  if (isTRUE(properties$has_direct_sample) && !is.function(direct_sample)) {
+    cli::cli_abort(
+      paste("Did not find a function 'direct_sample' within your model, but",
+            "your properties say that it should do"),
+      arg = "model", call = call)
+  }
+  if (!is.null(direct_sample) && !is.function(direct_sample)) {
+    cli::cli_abort(
+      "Expected 'model$direct_sample' to be a function if non-NULL",
+      arg = "model", call = call)
+  }
+  direct_sample
+}
```

Comment:
Could this repeated logic be implemented once and take expected property name etc as argument?

Reply:
yeah, I wondered about that. Will do